NeoML/src/TraditionalML/ProblemWrappers.cpp

```diff
@@ -199,6 +199,96 @@ CFloatVector CMultivariateRegressionOverBinaryClassification::GetValue( int inde
 	return classValues[classIndex];
 }
 
+/////////////////////////////////////////////////////////////////////////////////////////
+// CProblemNotNullWeightsView 
+
+CProblemNotNullWeightsView::CProblemNotNullWeightsView( const IProblem* _inner ) :
+	inner( _inner )
+{
+	NeoAssert( inner != 0 );
+
+	int originalVectorCount = inner->GetVectorCount();
+	if( originalVectorCount > 0 ) {
+		// fill nullWeightElementsMap, and adjust local MatrixDesc
+		viewMatrixDesc = _inner->GetMatrix();
+		// we are going to remap some elements, so create our own arrays of pointers
+		viewMatrixDesc.PointerB = static_cast<int*>( 
+			ALLOCATE_MEMORY( CurrentMemoryManager, originalVectorCount * sizeof( int ) ) );
+		viewMatrixDesc.PointerE = static_cast<int*>( 
+			ALLOCATE_MEMORY( CurrentMemoryManager, originalVectorCount * sizeof( int ) ) );
+
+		int nullWeightElmentsCount = 0;
+		for( int i = 0; i < originalVectorCount - nullWeightElmentsCount; ) {
+			int iScanned = i + nullWeightElmentsCount;
+			if( _inner->GetVectorWeight( iScanned ) == 0 ) {
+				++nullWeightElmentsCount;
+			} else {
+				viewMatrixDesc.PointerB[i] = _inner->GetMatrix().PointerB[iScanned];
+				viewMatrixDesc.PointerE[i] = _inner->GetMatrix().PointerE[iScanned];
+				notNullWeightElementsIndices.Add( iScanned );
+				++i;
+			}
+		}
+		viewMatrixDesc.Height -= nullWeightElmentsCount;
+	}
+}
+
+CProblemNotNullWeightsView::~CProblemNotNullWeightsView()
+{
+	if( inner->GetVectorCount() > 0 ) {
+		CurrentMemoryManager::Free( viewMatrixDesc.PointerB );
+		CurrentMemoryManager::Free( viewMatrixDesc.PointerE );
+	}
+}
+
+// Gets the number of classes
+inline int CProblemNotNullWeightsView::GetClassCount() const
+{
+	return inner->GetClassCount();
+}
+
+// Gets the number of features
+inline int CProblemNotNullWeightsView::GetFeatureCount() const
+{
+	return inner->GetFeatureCount();
+}
+
+// Indicates if the specified feature is discrete
+inline bool CProblemNotNullWeightsView::IsDiscreteFeature( int index ) const
+{
+	return inner->IsDiscreteFeature( index );
+}
+
+// Gets the number of vectors in the data set
+inline int CProblemNotNullWeightsView::GetVectorCount() const
+{
+	return viewMatrixDesc.Height;
+}
+
+// The correct class number for a vector with a given index in [0, GetClassCount())
+inline int CProblemNotNullWeightsView::GetClass( int index ) const
+{
+	return inner->GetClass( calculateOriginalIndex( index ) );
+}
+
+// Gets all vectors from the data set as a matrix
+inline CSparseFloatMatrixDesc CProblemNotNullWeightsView::GetMatrix() const
```

Comment:
Inline methods should be defined in header file.

Reply:
ok, moved to ProblemWrappers.inl